libraries/AP_RAMTRON/AP_RAMTRON.cpp

```diff
@@ -5,123 +5,203 @@
+        status_register &= ~kBlockProtect;
+        status_register |= kWriteEnableLatch; // Ensure Write Enable is not overwritten
+        if (!dev->write_register(RAMTRON_WRSR, status_register)) {
+            hal.console->printf("Failed to disable RAMTRON BlockProtect!");
+            return false;
         }
     }
-    hal.console->printf("Unknown RAMTRON manufacturer=%02x memory=%02x id1=%02x id2=%02x\n",
-                        rdid.manufacturer[0], rdid.memory, rdid.id1, rdid.id2);
-    return false;
+
+    return true;
 }
 
-/*
-  send a command and offset
- */
-void AP_RAMTRON::send_offset(uint8_t cmd, uint32_t offset)
+bool AP_RAMTRON::_fill_cmd_buffer(uint8_t cmdBuffer[], uint32_t const kCmdBufferSz, uint8_t const cmd, uint32_t addr)
 {
-    if (ramtron_ids[id].addrlen == 3) {
-        uint8_t b[4] = { cmd, uint8_t((offset>>16)&0xFF), uint8_t((offset>>8)&0xFF), uint8_t(offset&0xFF) };
-        dev->transfer(b, sizeof(b), nullptr, 0);
-    } else /* len 2 */ {
-        uint8_t b[3] = { cmd, uint8_t((offset>>8)&0xFF), uint8_t(offset&0xFF) };
-        dev->transfer(b, sizeof(b), nullptr, 0);
+    if (kCmdBufferSz == 4) {
+        cmdBuffer[1] = uint8_t((addr >> 16) & 0xFF);
+        cmdBuffer[2] = uint8_t((addr >>  8) & 0xFF);
+        cmdBuffer[3] = uint8_t((addr >>  0) & 0xFF);
+    } else if (kCmdBufferSz == 3) {
+        cmdBuffer[1] = uint8_t((addr >>  8) & 0xFF);
+        cmdBuffer[2] = uint8_t((addr >>  0) & 0xFF);
+    } else {
+        return false;
     }
+    cmdBuffer[0] = cmd;
+    return true;
 }
 
 // read from device
-bool AP_RAMTRON::read(uint32_t offset, uint8_t *buf, uint32_t size)
+uint32_t AP_RAMTRON::read(uint32_t offset, uint8_t * const buf, uint32_t size)
 {
-    const uint8_t maxread = 128;
-    while (size > maxread) {
-        if (!read(offset, buf, maxread)) {
-            return false;
-        }
-        offset += maxread;
-        buf += maxread;
-        size -= maxread;
+    // Don't allow reads outside of the FRAM memory.
+    // NOTE: The FRAM devices will wrap back to address 0x0000 if they read past
+    // the end of their internal memory, so while we'll get data back, it won't
+    // be what we expect.
+    if ((size > get_size()) ||
+        (offset > (get_size() - size))) {
+        return 0;
     }
 
-    WITH_SEMAPHORE(dev->get_semaphore());
+    uint32_t const kMaxReadSz = 128;
+    uint32_t numRead = 0;
 
-    dev->set_chip_select(true);
+    while (size > 0) {
+        uint32_t const kCmdBufferSz = ramtron_ids[id].addrlen + 1;
+        uint8_t cmdBuffer[kCmdBufferSz];
+        if (!_fill_cmd_buffer(cmdBuffer, kCmdBufferSz, RAMTRON_READ, (offset + numRead))) {
+            break;
+        } else {
+            WITH_SEMAPHORE(dev->get_semaphore());
 
-    send_offset(RAMTRON_READ, offset);
-    
-    // get data
-    dev->transfer(nullptr, 0, buf, size);
+            uint32_t const kReadSz = MIN(size, kMaxReadSz);
+            bool ok = dev->set_chip_select(true);
```

Comment:
transfer() already handles chip select, so is only needed when doing multiple operations within the one CS op

Reply:
Removed the unnecessary calls to `set_chip_select()`.